### PR TITLE
Implement SE-208 System Library Targets

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
<rdar://problem/39418939> [SR-7434]: Implement SE-208 Package Manager System Library Targets

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@sync create

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/PackageGraphLoader.swift [hunk 1]
```diff
@@ -252,10 +252,15 @@ private func createResolvedPackages(
         // The diagnostics location for this package.
         let diagnosticLocation = { PackageLocation.Local(name: package.name, packagePath: package.path) }
 
-        // Get all the system module dependencies in this package.
-        let systemModulesDeps = packageBuilder.dependencies
+        // Get all implicit system library dependencies in this package.
+        let implicitSystemTargetDeps = packageBuilder.dependencies
             .flatMap({ $0.targets })
-            .filter({ $0.target.type == .systemModule })
+            .filter({
```
Comment (hartbit) [inline Sources/PackageGraph/PackageGraphLoader.swift]: Could we simplify this to `($0.target as? SystemLibraryTarget)?.isImplicit ?? false`

File: Sources/PackageLoading/PackageBuilder.swift [hunk 2]
```diff
@@ -691,6 +711,19 @@ public final class PackageBuilder {
...
     ) throws -> Target? {
 
+        // Create system library target.
+        if potentialModule.type == .system {
+            let moduleMapPath = potentialModule.path.appending(component: moduleMapFilename)
+            guard fileSystem.isFile(moduleMapPath) else {
+                return nil
+            }
+            return SystemLibraryTarget(
+                    name: potentialModule.name,
```
Comment (hartbit) [inline Sources/PackageLoading/PackageBuilder.swift]: nitpick: indentation

File: Sources/PackageLoading/Diagnostics.swift [hunk 3]
```diff
@@ -76,4 +76,41 @@ public enum PackageBuilderDiagnostics {
...
+            }
+        )
+
+        let targets: [String]
+    }
+
+    struct SystemPackageProductValidationDiagnostic: DiagnosticData {
+        static let id = DiagnosticID(
+            type: SystemPackageProductValidationDiagnostic.self,
+            name: "org.swift.diags.pkg-builder.sys-pkg-product-validatoin",
```
Comment (hartbit) [inline Sources/PackageLoading/Diagnostics.swift]: typo: validation

File: Sources/PackageLoading/Diagnostics.swift [hunk 4]
```diff
@@ -76,4 +76,41 @@ public enum PackageBuilderDiagnostics {
...
+            defaultBehavior: .warning,
+            description: {
+                $0 <<< "Ignoring declared target(s)" <<< { "'\($0.targets.joined(separator: ", "))'" } <<< "in the system package"
+            }
+        )
+
+        let targets: [String]
+    }
+
+    struct SystemPackageProductValidationDiagnostic: DiagnosticData {
```
Comment (hartbit) [inline Sources/PackageLoading/Diagnostics.swift]: I don't understand this diagnostic. Can you explain?

File: Sources/PackageLoading/PackageBuilder.swift [hunk 5]
```diff
@@ -1054,6 +1087,19 @@ public final class PackageBuilder {
                     case nil: type = .library(.automatic)
                     }
                     let targets = try modulesFrom(targetNames: p.targets, product: p.name)
+
+                    // Peform special validations if this product is exporting
+                    // a system library target.
+                    if targets.contains(where: { $0 is SystemLibraryTarget }) {
+                        if type != .library(.automatic) || targets.count != 1 {
```
Comment (hartbit) [inline Sources/PackageLoading/PackageBuilder.swift]: Minor comment: I find the double-ifs hard to read (perhaps that's just me).

File: Sources/PackageLoading/Diagnostics.swift (hunk 4 above)
Comment (aciidb0mb3r) [reply Sources/PackageLoading/Diagnostics.swift]: System library products shouldn't have a type (automatic/static/dynamic) and should contains only one target.

File: Sources/PackageGraph/PackageGraphLoader.swift (hunk 1 above)
Comment (aciidb0mb3r) [reply Sources/PackageGraph/PackageGraphLoader.swift]: I wouldn't call one or the other a simplified version, heh.

**Pull request merged**